File: python/tempmail_wrapper/providers/dropmail.py
"""dropmail.me provider implementation (GraphQL)."""

from __future__ import annotations

import logging
import warnings
from collections.abc import Callable
from datetime import datetime

import requests

from ..base import TempMailProvider
from ..exceptions import NotFoundError, TempMailError
from ..http_client import HttpClient
from ..models import Message, MessageDetail

logger = logging.getLogger(__name__)

PADDLE_OCR_URL = "https://mamamacjdjj-padle-predict.hf.space/predict"


def paddle_ocr_solver(img_bytes: bytes) -> str | None:
    """Solve captcha via PaddleOCR HuggingFace space (3 retries)."""
    for attempt in range(3):
        try:
            ocr_resp = requests.post(
                PADDLE_OCR_URL,
                files={"file": ("cap.png", img_bytes, "image/png")},
                timeout=20,
            )
            ocr_data = ocr_resp.json()
            results = ocr_data.get("results", [])
            if results:
                best = results[0]
                confidence = best.get("confidence", 0)
                if confidence >= 0.7:
                    text = best.get("text", "").strip()
                    if text:
                        return text
                logger.debug(
                    "Dropmail: OCR attempt %d confidence %.3f < 0.7, retrying",
                    attempt + 1, confidence,
                )
        except Exception as exc:
            logger.debug("Dropmail: OCR attempt %d failed: %s", attempt + 1, exc)
    return None


class DropmailProvider(TempMailProvider):
    """Provider for dropmail.me — GraphQL API, no auth required."""

    TOKEN_URL = "https://dropmail.me/api/token/generate"

    def __init__(
        self,
        proxies: list[str] | None = None,
        random_ua: bool = True,
        solvers: list[Callable[[bytes], str | None]] | None = None,
        **kwargs,
    ) -> None:
        self._http = HttpClient(proxies=proxies, random_ua=random_ua, use_cookies=True)
        self._solvers = solvers
        self._token: str | None = None
        self._graphql_url: str | None = None
        self._session_id: str | None = None
        self._address_id: str | None = None
        self._email: str | None = None

    @property
    def name(self) -> str:
        return "dropmail"

    def _gql(self, query: str, variables: dict | None = None) -> dict:
        if not self._token:
            raise TempMailError("No token. Call generate_email() first.", self.name)
        payload: dict = {"query": query}
        if variables:
            payload["variables"] = variables
        resp = self._http.post(self._graphql_url, json=payload, timeout=30)  # type: ignore[arg-type]
        if resp.status_code != 200:
            raise TempMailError(f"GraphQL error: {resp.status_code} {resp.text}", self.name)
        data = resp.json()
        if "errors" in data:
            raise TempMailError(f"GraphQL errors: {data['errors']}", self.name)
        return data.get("data", {})

    def _solve_captcha_and_get_token(self, captcha: dict) -> str | None:
        """Download captcha image (same session), solve via PaddleOCR, submit solution.
        Returns 90d token on success, None on failure (caller falls back to 1d).
        """
        v = captcha.get("v", "3")
        nonce = captcha.get("nonce", "")
        key = captcha.get("key", "")
        sig = captcha.get("_sig", "")

        img_url = (
            f"https://dropmail.me/captcha/image.png"
            f"?_r=0&v={v}&nonce={nonce}&key={key}&_sig={sig}"
        )

        # Step 2: download image — SAME session (cookies matter)
        img_resp = self._http.get(img_url, timeout=15)
        if img_resp.status_code != 200:
            logger.warning("Dropmail: captcha image download failed (%s)", img_resp.status_code)
            return None
        img_bytes = img_resp.content

        # Step 3: solve via solver chain
        text = None
        solvers = self._solvers if self._solvers is not None else [paddle_ocr_solver]
        for solver in solvers:
            try:
                result = solver(img_bytes)
                if result and result.strip():
                    text = result.strip()
                    break
            except Exception as exc:
                logger.debug("Dropmail: solver error: %s", exc)

        if not text:
            logger.warning("Dropmail: all solvers failed, falling back to 1d token")
            return None

        # Step 4: submit solution — SAME session
        solution_resp = self._http.post(
            "https://dropmail.me/captcha/solution",
            data={"response": text, "v": v, "nonce": nonce, "key": key, "_sig": sig},
            timeout=15,
        )
        try:
            sol_data = solution_resp.json()
        except Exception:
            logger.warning("Dropmail: captcha solution response not JSON, falling back")
            return None

        if sol_data.get("result") != "correct":
            logger.warning(
                "Dropmail: captcha solution rejected (%s), falling back to 1d token",
                sol_data,
            )
            return None

        # Step 5: retry token generation with 90d — SAME session
        token_resp = self._http.post(
            self.TOKEN_URL,
            json={"type": "af", "lifetime": "90d"},
            timeout=30,
        )
        if token_resp.status_code != 200:
            logger.warning(
                "Dropmail: 90d token generation failed (%s) after captcha solve, falling back",
                token_resp.status_code,
            )
            return None

        token = token_resp.json().get("token")
        if not token:
            logger.warning("Dropmail: no token in 90d response, falling back")
            return None

        return token

    def generate_email(self) -> str:
        # Step 1: try 1d token (fast, no captcha)
        resp = self._http.post(
            self.TOKEN_URL,
            json={"type": "af", "lifetime": "1d"},
            timeout=30,
        )

        if resp.status_code == 200:
            data = resp.json()
            self._token = data.get("token")
            if not self._token:
                raise TempMailError("No token in response", self.name)
        elif resp.status_code == 402:
            # Captcha required — try 90d flow
            try:
                captcha_data = resp.json().get("captcha", {})
                self._token = self._solve_captcha_and_get_token(captcha_data)
            except Exception as exc:
                logger.warning("Dropmail: captcha flow error: %s", exc)
                self._token = None

            if not self._token:
                # Fallback: request 1d token again on fresh session
                warnings.warn(
                    "Dropmail: captcha solve failed, retrying with 1d token",
                    stacklevel=2,
                )
                resp2 = self._http.post(
                    self.TOKEN_URL,
                    json={"type": "af", "lifetime": "1d"},
                    timeout=30,
                )
                if resp2.status_code != 200:
                    raise TempMailError(
                        f"Token generation failed: {resp2.status_code}", self.name
                    )
                self._token = resp2.json().get("token")
                if not self._token:
                    raise TempMailError("No token in response", self.name)
        else:
            raise TempMailError(f"Token generation failed: {resp.status_code}", self.name)

        self._graphql_url = f"https://dropmail.me/api/graphql/{self._token}"

        mutation = """
            mutation {
                introduceSession {
                    id
                    addresses {
                        id
                        address
                        restoreKey
                    }
                }
            }
        """
        data = self._gql(mutation)
        session = data.get("introduceSession")
        if not session:
            raise TempMailError("Failed to create session", self.name)

        self._session_id = session["id"]
        addresses = session.get("addresses", [])
        if not addresses:
            raise TempMailError("No address generated", self.name)

        addr = addresses[0]
        self._address_id = addr["id"]
        self._email = addr["address"]
        return self._email

    def get_inbox(self, email: str) -> list[Message]:
        query = """
            query($sid: ID!) {
                session(id: $sid) {
                    mails {
                        id
                        fromAddr
                        headerSubject
                        receivedAt
                    }
                }
            }
        """
        data = self._gql(query, {"sid": self._session_id})
        session_data = data.get("session", {})
        messages = []
        for mail in session_data.get("mails", []):
            received = mail.get("receivedAt", "2000-01-01T00:00:00Z")
            if received.endswith("Z"):
                received = received[:-1] + "+00:00"
            try:
                date = datetime.fromisoformat(received)
            except ValueError:
                date = datetime.now()
            messages.append(
                Message(
                    id=mail["id"],
                    sender=mail.get("fromAddr", ""),
                    subject=mail.get("headerSubject", ""),
                    date=date,
                )
            )
        return messages

    def read_message(self, message_id: str) -> MessageDetail:
        query = """
            query($sid: ID!) {
                session(id: $sid) {
                    mails {
                        id
                        fromAddr
                        headerSubject
                        receivedAt
                        text
                        html
                        attachments {
                            name
                            mime
                            rawSize
                        }
                    }
                }
            }
        """
        data = self._gql(query, {"sid": self._session_id})
        session_data = data.get("session", {})

        for mail in session_data.get("mails", []):
            if mail["id"] == message_id:
                received = mail.get("receivedAt", "2000-01-01T00:00:00Z")
                if received.endswith("Z"):
                    received = received[:-1] + "+00:00"
                try:
                    date = datetime.fromisoformat(received)
                except ValueError:
                    date = datetime.now()

                attachments = [
                    {
                        "filename": att.get("name", ""),
                        "content_type": att.get("mime", ""),
                        "size": att.get("rawSize", 0),
                    }
                    for att in mail.get("attachments", [])
                ]

                return MessageDetail(
                    id=mail["id"],
                    sender=mail.get("fromAddr", ""),
                    subject=mail.get("headerSubject", ""),
                    date=date,
                    body_text=mail.get("text", ""),
                    body_html=mail.get("html", ""),
                    attachments=attachments,
                )

        raise NotFoundError(f"Message {message_id} not found", self.name)

    def delete_email(self, email: str) -> bool:
        if not self._address_id:
            return True

        mutation = """
            mutation($input: DeleteAddressInput!) {
                deleteAddress(input: $input) {
                    id
                }
            }
        """
        try:
            self._gql(mutation, {"input": {"addressId": self._address_id}})
        except TempMailError:
            pass

        self._token = None
        self._graphql_url = None
        self._session_id = None
        self._address_id = None
        self._email = None
        return True

File: go/providers/dropmail.go
package providers

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

const dropmailPaddleOCRURL = "https://mamamacjdjj-padle-predict.hf.space/predict"

// PaddleOCRSolver solves a captcha image via PaddleOCR HuggingFace space (3 retries).
func PaddleOCRSolver(imgBytes []byte) string {
	for attempt := 0; attempt < 3; attempt++ {
		var buf bytes.Buffer
		mw := multipart.NewWriter(&buf)
		part, _ := mw.CreateFormFile("file", "cap.png")
		part.Write(imgBytes)
		mw.Close()

		ocrResp, err := http.Post(dropmailPaddleOCRURL, mw.FormDataContentType(), &buf) //nolint:noctx
		if err != nil {
			continue
		}
		var ocrData struct {
			Results []struct {
				Text       string  `json:"text"`
				Confidence float64 `json:"confidence"`
			} `json:"results"`
		}
		json.NewDecoder(ocrResp.Body).Decode(&ocrData)
		ocrResp.Body.Close()

		if len(ocrData.Results) > 0 && ocrData.Results[0].Confidence >= 0.7 {
			if text := strings.TrimSpace(ocrData.Results[0].Text); text != "" {
				return text
			}
		}
	}
	return ""
}

type Dropmail struct {
	client         *HttpClient
	sessions       map[string]*dropmailSession
	mu             sync.Mutex
	CaptchaSolvers []func(imgBytes []byte) string
}

type dropmailSession struct {
	Token     string
	SessionID string
	AddressID string
}

func NewDropmail(config map[string]string) *Dropmail {
	proxies := ParseProxies(config["proxies"])
	randomUA := config["randomUa"] == "true"

	return &Dropmail{
		client:   NewHttpClient(proxies, randomUA, false),
		sessions: make(map[string]*dropmailSession),
	}
}

// dropmailRawPost performs a POST using the shared HttpClient cookie jar,
// returning the raw http.Response (caller must close Body).
func (d *Dropmail) dropmailRawPost(url string, contentType string, body []byte) (*http.Response, error) {
	req, _ := http.NewRequest(http.MethodPost, url, bytes.NewReader(body))
	req.Header.Set("Content-Type", contentType)
	// Use GetClient so cookies are shared with the jar.
	client := d.client.GetClient()
	return client.Do(req)
}

// dropmailRawGet performs a GET using the shared HttpClient cookie jar.
func (d *Dropmail) dropmailRawGet(rawURL string) (*http.Response, error) {
	req, _ := http.NewRequest(http.MethodGet, rawURL, nil)
	client := d.client.GetClient()
	return client.Do(req)
}

// solveCaptchaAndGetToken runs the 5-step 90d captcha flow.
// Returns the token on success, "" on failure (caller falls back to 1d).
func (d *Dropmail) solveCaptchaAndGetToken(captcha map[string]interface{}) string {
	v, _ := captcha["v"].(string)
	nonce, _ := captcha["nonce"].(string)
	key, _ := captcha["key"].(string)
	sig, _ := captcha["_sig"].(string)

	// Step 2: download captcha image — same cookie jar
	imgURL := fmt.Sprintf(
		"https://dropmail.me/captcha/image.png?_r=0&v=%s&nonce=%s&key=%s&_sig=%s",
		url.QueryEscape(v), url.QueryEscape(nonce), url.QueryEscape(key), url.QueryEscape(sig),
	)
	imgResp, err := d.dropmailRawGet(imgURL)
	if err != nil || imgResp.StatusCode != 200 {
		if imgResp != nil {
			imgResp.Body.Close()
		}
		return ""
	}
	imgBytes, err := io.ReadAll(imgResp.Body)
	imgResp.Body.Close()
	if err != nil {
		return ""
	}

	// Step 3: solve via solver chain
	var ocrText string
	solvers := d.CaptchaSolvers
	if len(solvers) == 0 {
		solvers = []func([]byte) string{PaddleOCRSolver}
	}
	for _, solver := range solvers {
		result := strings.TrimSpace(solver(imgBytes))
		if result != "" {
			ocrText = result
			break
		}
	}
	if ocrText == "" {
		return ""
	}

	// Step 4: submit solution — same cookie jar, form-encoded
	formVals := url.Values{
		"response": {ocrText},
		"v":        {v},
		"nonce":    {nonce},
		"key":      {key},
		"_sig":     {sig},
	}
	solBody := []byte(formVals.Encode())
	solResp, err := d.dropmailRawPost(
		"https://dropmail.me/captcha/solution",
		"application/x-www-form-urlencoded",
		solBody,
	)
	if err != nil {
		return ""
	}
	var solData struct {
		Result string `json:"result"`
	}
	json.NewDecoder(solResp.Body).Decode(&solData)
	solResp.Body.Close()
	if solData.Result != "correct" {
		return ""
	}

	// Step 5: retry token generation with 90d — same cookie jar
	tokenBody, _ := json.Marshal(map[string]string{"type": "af", "lifetime": "90d"})
	tokenResp, err := d.dropmailRawPost(
		"https://dropmail.me/api/token/generate",
		"application/json",
		tokenBody,
	)
	if err != nil || tokenResp.StatusCode != 200 {
		if tokenResp != nil {
			tokenResp.Body.Close()
		}
		return ""
	}
	var tokenData struct {
		Token string `json:"token"`
	}
	json.NewDecoder(tokenResp.Body).Decode(&tokenData)
	tokenResp.Body.Close()
	return tokenData.Token
}

func (d *Dropmail) GenerateEmail() (string, error) {
	// Step 1: try 1d token (fast, no captcha)
	reqBody, _ := json.Marshal(map[string]string{"type": "af", "lifetime": "1d"})
	req, _ := http.NewRequest(http.MethodPost, "https://dropmail.me/api/token/generate", bytes.NewReader(reqBody))
	req.Header.Set("Content-Type", "application/json")

	// Use GetClient so the response is visible and cookies are stored.
	rawClient := d.client.GetClient()
	resp, err := rawClient.Do(req)
	if err != nil {
		return "", err
	}
	respBody, _ := io.ReadAll(resp.Body)
	resp.Body.Close()

	var token string

	if resp.StatusCode == 200 {
		var tokenResp struct {
			Token string `json:"token"`
		}
		if err := json.Unmarshal(respBody, &tokenResp); err != nil || tokenResp.Token == "" {
			return "", &TempMailError{Provider: "dropmail", Message: "failed to get token"}
		}
		token = tokenResp.Token
	} else if resp.StatusCode == 402 {
		// Captcha required — attempt 90d flow
		var capResp struct {
			Captcha map[string]interface{} `json:"captcha"`
		}
		json.Unmarshal(respBody, &capResp)
		token = d.solveCaptchaAndGetToken(capResp.Captcha)

		if token == "" {
			// Fallback: retry 1d
			req2, _ := http.NewRequest(http.MethodPost, "https://dropmail.me/api/token/generate", bytes.NewReader(reqBody))
			req2.Header.Set("Content-Type", "application/json")
			resp2, err := d.client.RawDo(req2, reqBody)
			if err != nil {
				return "", err
			}
			var tokenResp2 struct {
				Token string `json:"token"`
			}
			if err := json.Unmarshal(resp2, &tokenResp2); err != nil || tokenResp2.Token == "" {
				return "", &TempMailError{Provider: "dropmail", Message: "failed to get token (1d fallback)"}
			}
			token = tokenResp2.Token
		}
	} else {
		return "", &TempMailError{Provider: "dropmail", Message: fmt.Sprintf("token generation failed: HTTP %d", resp.StatusCode)}
	}

	graphql := `{"query": "mutation { introduceSession { id addresses { id address restoreKey } } }"}`
	req2, _ := http.NewRequest(http.MethodPost, fmt.Sprintf("https://dropmail.me/api/graphql/%s", token), bytes.NewBufferString(graphql))
	req2.Header.Set("Content-Type", "application/json")
	respBody2, err := d.client.RawDo(req2, []byte(graphql))
	if err != nil {
		return "", err
	}

	var sessionResp struct {
		Data struct {
			IntroduceSession struct {
				ID        string `json:"id"`
				Addresses []struct {
					ID      string `json:"id"`
					Address string `json:"address"`
				} `json:"addresses"`
			} `json:"introduceSession"`
		} `json:"data"`
	}
	if err := json.Unmarshal(respBody2, &sessionResp); err != nil {
		return "", &TempMailError{Provider: "dropmail", Message: "failed to parse session"}
	}

	addrs := sessionResp.Data.IntroduceSession.Addresses
	if len(addrs) == 0 {
		return "", &TempMailError{Provider: "dropmail", Message: "no address created"}
	}

	email := addrs[0].Address
	sess := &dropmailSession{
		Token:     token,
		SessionID: sessionResp.Data.IntroduceSession.ID,
		AddressID: addrs[0].ID,
	}

	d.mu.Lock()
	d.sessions[email] = sess
	d.mu.Unlock()

	return email, nil
}

func (d *Dropmail) GetInbox(email string) ([]Message, error) {
	d.mu.Lock()
	sess := d.sessions[email]
	d.mu.Unlock()
	if sess == nil {
		return nil, &TempMailError{Provider: "dropmail", Message: "unknown email"}
	}

	query := fmt.Sprintf(`{"query": "query { session(id: \"%s\") { mails { id fromAddr headerSubject receivedAt text html attachments { id name mime rawSize } } } }"}`, sess.SessionID)
	req, _ := http.NewRequest(http.MethodPost, fmt.Sprintf("https://dropmail.me/api/graphql/%s", sess.Token), bytes.NewBufferString(query))
	req.Header.Set("Content-Type", "application/json")
	body, err := d.client.RawDo(req, []byte(query))
	if err != nil {
		return nil, err
	}

	var inboxResp struct {
		Data struct {
			Session struct {
				Mails []struct {
					ID            string `json:"id"`
					FromAddr      string `json:"fromAddr"`
					HeaderSubject string `json:"headerSubject"`
					ReceivedAt    string `json:"receivedAt"`
				} `json:"mails"`
			} `json:"session"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &inboxResp); err != nil {
		return nil, &TempMailError{Provider: "dropmail", Message: "parse error"}
	}

	var msgs []Message
	for _, m := range inboxResp.Data.Session.Mails {
		date, _ := time.Parse(time.RFC3339, m.ReceivedAt)
		msgs = append(msgs, Message{
			ID:      fmt.Sprintf("%s|%s", email, m.ID),
			Sender:  m.FromAddr,
			Subject: m.HeaderSubject,
			Date:    date,
		})
	}
	return msgs, nil
}

func (d *Dropmail) ReadMessage(messageID string) (*MessageDetail, error) {
	parts := strings.SplitN(messageID, "|", 2)
	if len(parts) != 2 {
		return nil, &TempMailError{Provider: "dropmail", Message: "invalid message ID"}
	}
	email, realID := parts[0], parts[1]

	d.mu.Lock()
	sess := d.sessions[email]
	d.mu.Unlock()
	if sess == nil {
		return nil, &TempMailError{Provider: "dropmail", Message: "unknown email"}
	}

	query := fmt.Sprintf(`{"query": "query { session(id: \"%s\") { mails { id fromAddr headerSubject receivedAt text html attachments { id name mime rawSize } } } }"}`, sess.SessionID)
	req, _ := http.NewRequest(http.MethodPost, fmt.Sprintf("https://dropmail.me/api/graphql/%s", sess.Token), bytes.NewBufferString(query))
	req.Header.Set("Content-Type", "application/json")
	body, err := d.client.RawDo(req, []byte(query))
	if err != nil {
		return nil, err
	}

	var inboxResp struct {
		Data struct {
			Session struct {
				Mails []struct {
					ID            string           `json:"id"`
					FromAddr      string           `json:"fromAddr"`
					HeaderSubject string           `json:"headerSubject"`
					ReceivedAt    string           `json:"receivedAt"`
					Text          string           `json:"text"`
					HTML          string           `json:"html"`
					Attachments   []map[string]any `json:"attachments"`
				} `json:"mails"`
			} `json:"session"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &inboxResp); err != nil {
		return nil, &TempMailError{Provider: "dropmail", Message: "parse error"}
	}

	for _, m := range inboxResp.Data.Session.Mails {
		if m.ID == realID {
			date, _ := time.Parse(time.RFC3339, m.ReceivedAt)
			return &MessageDetail{
				Message: Message{
					ID:      messageID,
					Sender:  m.FromAddr,
					Subject: m.HeaderSubject,
					Date:    date,
				},
				BodyText:    m.Text,
				BodyHTML:    m.HTML,
				Attachments: m.Attachments,
			}, nil
		}
	}
	return nil, &NotFoundError{Provider: "dropmail", Resource: "message"}
}

func (d *Dropmail) DeleteEmail(email string) (bool, error) {
	d.mu.Lock()
	sess := d.sessions[email]
	if sess != nil {
		delete(d.sessions, email)
	}
	d.mu.Unlock()

	if sess == nil {
		return true, nil
	}

	query := fmt.Sprintf(`{"query": "mutation { deleteAddress(input: { addressId: \"%s\" }) { id } }"}`, sess.AddressID)
	req, _ := http.NewRequest(http.MethodPost, fmt.Sprintf("https://dropmail.me/api/graphql/%s", sess.Token), bytes.NewBufferString(query))
	req.Header.Set("Content-Type", "application/json")
	_, err := d.client.RawDo(req, []byte(query))
	if err != nil {
		return false, err
	}
	return true, nil
}

func (d *Dropmail) WaitForEmail(email string, timeout time.Duration, interval time.Duration) (*Message, error) {
	return waitForEmail(d, email, timeout, interval)
}
